sdk/trace/src/main/scala/org/typelevel/otel4s/sdk/trace/SpanProcessor.scala

```diff
@@ -0,0 +1,161 @@
+    */
+  def isStartRequired: Boolean
+
+  /** Called when a span is ended, if the `span.isRecording` returns true.
+    *
+    * This method is called synchronously on the execution thread, should not
+    * throw or block the execution thread.
+    *
+    * @param span
+    *   the ended span
+    */
+  def onEnd(span: SpanData): F[Unit]
+
+  /** Whether the [[SpanProcessor]] requires end events.
+    *
+    * If true, the [[onEnd]] will be called upon the end of a span.
+    */
+  def isEndRequired: Boolean
+
+  /** Processes all pending spans (if any).
+    */
+  def forceFlush: F[Unit]
+}
+
+object SpanProcessor {
+
+  /** Creates a [[SpanProcessor]] which delegates all processing to the
+    * processors in order.
+    */
+  def of[F[_]: Applicative: Parallel](
+      processors: SpanProcessor[F]*
+  ): SpanProcessor[F] =
+    if (processors.sizeIs == 1) processors.head
+    else processors.combineAll
+
+  /** Creates a no-op implementation of the [[SpanProcessor]].
+    *
+    * All export operations are no-op.
+    */
+  def noop[F[_]: Applicative]: SpanProcessor[F] =
+    new Noop
+
+  implicit def spanProcessorMonoid[F[_]: Applicative: Parallel]
+      : Monoid[SpanProcessor[F]] =
+    new Monoid[SpanProcessor[F]] {
+      val empty: SpanProcessor[F] =
+        noop[F]
+
+      def combine(x: SpanProcessor[F], y: SpanProcessor[F]): SpanProcessor[F] =
+        (x, y) match {
+          case (that, _: Noop[F]) =>
+            that
+          case (_: Noop[F], other) =>
+            other
+          case (that: Multi[F], other: Multi[F]) =>
+            Multi(that.processors.concatNel(other.processors))
+          case (that: Multi[F], other) =>
+            Multi(that.processors :+ other)
+          case (that, other: Multi[F]) =>
+            Multi(that :: other.processors)
+          case (that, other) =>
+            Multi(NonEmptyList.of(that, other))
+        }
+    }
+
+  private final class Noop[F[_]: Applicative] extends SpanProcessor[F] {
+    def isStartRequired: Boolean = false
+    def isEndRequired: Boolean = false
+
+    def onStart(parentCtx: Option[SpanContext], span: SpanRef[F]): F[Unit] =
+      Applicative[F].unit
+
+    def onEnd(span: SpanData): F[Unit] =
+      Applicative[F].unit
+
+    def forceFlush: F[Unit] =
+      Applicative[F].unit
+
+    override def toString: String = "SpanProcessor.Noop"
+  }
+
+  private final case class Multi[F[_]: Parallel](
+      processors: NonEmptyList[SpanProcessor[F]]
+  ) extends SpanProcessor[F] {
+    private val startOnly: List[SpanProcessor[F]] =
+      processors.filter(_.isStartRequired)
+
+    private val endOnly: List[SpanProcessor[F]] =
+      processors.filter(_.isEndRequired)
+
+    def isStartRequired: Boolean = startOnly.nonEmpty
+    def isEndRequired: Boolean = endOnly.nonEmpty
+
+    def onStart(parentCtx: Option[SpanContext], span: SpanRef[F]): F[Unit] =
+      startOnly.parTraverse_(_.onStart(parentCtx, span))
```

Comment:
Usual issue here about failure and short-circuiting.

Reply:
.attempt should be enough